document/create.go

```diff
@@ -75,94 +75,70 @@ func NewFromStruct(s interface{}) (Document, error) {
 		return nil, errors.New("expected struct or pointer to struct")
 	}
 
-	return structDocument{ref: ref}, nil
+	return newFromStruct(ref)
 }
 
-type structDocument struct {
-	ref reflect.Value
-}
-
-var _ Document = (*structDocument)(nil)
-
-func (s structDocument) Iterate(fn func(field string, value Value) error) error {
-	l := s.ref.NumField()
-
-	tp := s.ref.Type()
+func newFromStruct(ref reflect.Value) (Document, error) {
+	var fb FieldBuffer
+	l := ref.NumField()
+	tp := ref.Type()
 
 	for i := 0; i < l; i++ {
-		sf := tp.Field(i)
-		if sf.PkgPath != "" {
+		f := reflect.Indirect(ref.Field(i))
+		if !f.IsValid() {
 			continue
 		}
 
-		var name string
-		if gtag, ok := sf.Tag.Lookup("genji"); ok {
-			if gtag == "-" {
-				continue
+		sf := tp.Field(i)
+
+		if f.Kind() == reflect.Struct {
+			d, err := newFromStruct(f)
+			if err != nil {
+				return nil, err
 			}
 
-			name = gtag
-		} else {
-			name = strings.ToLower(sf.Name)
-		}
+			// Embedded struct.
+			if sf.Anonymous {
+				err := d.Iterate(func(field string, value Value) error {
+					fb.Add(field, value)
+					return nil
+				})
 
-		f := s.ref.Field(i)
+				if err != nil {
+					return nil, err
+				}
 
-		v, err := NewValue(f.Interface())
-		if err != nil {
-			if err.(*ErrUnsupportedType) != nil {
-				continue
+			} else {
+				fb.Add(strings.ToLower(sf.Name), NewDocumentValue(d))
 			}
-			return err
-		}
 
-		err = fn(name, v)
-		if err != nil {
-			return err
+			continue
 		}
-	}
 
-	return nil
-}
-
-func (s structDocument) GetByField(field string) (Value, error) {
-	tp := s.ref.Type()
-
-	var sf reflect.StructField
-	var ok bool
-
-	ln := tp.NumField()
-	for i := 0; i < ln; i++ {
-		sf = tp.Field(i)
-		if gtag, found := sf.Tag.Lookup("genji"); found && gtag == field {
-			ok = true
-			break
+		// Ignore unexported field.
+		if sf.PkgPath != "" {
+			continue
 		}
-		if strings.ToLower(sf.Name) == field {
-			ok = true
-			break
+
+		field := strings.ToLower(sf.Name)
+		if gtag, ok := sf.Tag.Lookup("genji"); ok && gtag != "-" {
```

Comment:
I am wondering if we should support the `omitempty` tag. What do you guys think? cc @asdine

Reply:
I think it can be useful if for example the user wants to add a field to the document only if it's not empty otherwise it is skipped.

```go
type Foo struct {
	Bar string
	Baz string `genji:"omitempty"`
}

foo1 := Foo{Bar:"bar", Baz:"baz"}
// {Bar:"bar", Baz:"baz"}

foo2 := Foo{Bar:"bar"}
// {Bar:"bar"} not {Bar:"bar", Baz:""}

```

Reply:
It make sense, just be sure is not a confusion point for the user. He should be sure that using `omitempty` disallow default value.

Reply:
That's a good question, but since it's not directly related to this PR can you open an issue about this @yaziine?

Reply:
Could I mark this as resolve ?